File: app/src/main/java/com/seanmiller/alpacadashboard/RecyclerViewAdapterPositions.java
package com.seanmiller.alpacadashboard;

import android.content.Context;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.card.MaterialCardView;
import com.mashape.unirest.http.Unirest;
import com.mashape.unirest.http.exceptions.UnirestException;

import net.jacobpeterson.alpaca.AlpacaAPI;
import net.jacobpeterson.alpaca.rest.exception.AlpacaAPIRequestException;
import net.jacobpeterson.domain.alpaca.calendar.Calendar;
import net.jacobpeterson.domain.alpaca.position.Position;
import net.jacobpeterson.domain.polygon.aggregates.Aggregate;
import net.jacobpeterson.domain.polygon.lastquote.LastQuoteResponse;
import net.jacobpeterson.polygon.PolygonAPI;
import net.jacobpeterson.polygon.enums.Timespan;
import net.jacobpeterson.polygon.rest.exception.PolygonAPIRequestException;

import org.json.JSONException;
import org.json.JSONObject;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

public class RecyclerViewAdapterPositions extends RecyclerView.Adapter<RecyclerViewAdapterPositions.ViewHolder> {

    private static List<String> mData;
    private final LayoutInflater mInflater;
    private ItemClickListener mClickListener;

    // data is passed into the constructor
    RecyclerViewAdapterPositions(Context context, List<String> data) {
        this.mInflater = LayoutInflater.from(context);
        mData = data;
    }


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.recyclerview_position, parent, false);
        return new ViewHolder(view);
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    public void onBindViewHolder(@NonNull RecyclerViewAdapterPositions.ViewHolder holder, int position) {

        MainActivity mainActivity = new MainActivity();

        String stockName = mData.get(position);
        holder.stock_name.setText(stockName);
        SharedPreferencesManager prefs = new SharedPreferencesManager(mInflater.getContext());

        Thread thread = new Thread(() -> {

            PolygonAPI polygonAPI = new PolygonAPI(prefs.retrieveString("polygon_id", "NULL"));
            AlpacaAPI alpacaAPI = new AlpacaAPI(prefs.retrieveString("auth_token", "NULL"));

            // Fetch last open day's information
            ArrayList<Calendar> calendar = null;
            try {
                calendar = alpacaAPI.getCalendar(LocalDate.now().minusWeeks(1), LocalDate.now());
            } catch (AlpacaAPIRequestException e) {
                e.printStackTrace();
            }
            assert calendar != null;
            LocalDate lastOpenDate = LocalDate.parse(calendar.get(calendar.size() - 2).getDate());
            if (LocalTime.of(Integer.parseInt(calendar.get(calendar.size() - 2).getOpen().substring(0, 2)),
                    Integer.parseInt(calendar.get(calendar.size() - 2).getOpen().substring(3, 5))).compareTo(LocalTime.now()) > 0) {
                lastOpenDate = LocalDate.parse(calendar.get(calendar.size() - 3).getDate());
            }

            // Get Last value
            float close = 0;
            float closeCurr = 0;
            try {

                // Fetch Daily Open Close endpoint
                // https://api.polygon.io/v1/open-close/AAPL/2020-10-14?apiKey=
                JSONObject nodeHttpResponse = null;
                try {
                    nodeHttpResponse = Unirest.get("https://api.polygon.io/v1/open-close/" + mData.get(position) + "/" + lastOpenDate + "?apiKey=" + prefs.retrieveString("polygon_id", "NULL")).asJson().getBody().getObject();

                } catch (UnirestException e) {
                    e.printStackTrace();
                }

                if (nodeHttpResponse != null) {
                    close = Float.parseFloat(nodeHttpResponse.get("close").toString());
                }

            } catch (JSONException e) {
                e.printStackTrace();
            }

//            while (true) {

            // Get Amount of shares owned
            Position shrOwned = null;
            try {
                shrOwned = alpacaAPI.getOpenPositionBySymbol(mData.get(position));
            } catch (AlpacaAPIRequestException e) {
                e.printStackTrace();
            }

            // Set shares owned
            Position finalShrOwned = shrOwned;
            mainActivity.runOnUiThread(() -> {
                if (finalShrOwned.getQty() != null) {

                    if (finalShrOwned.getQty().equals("1")) {
                        holder.sharesOwned.setText(String.format("%s share owned", finalShrOwned.getQty()));
                    } else {
                        holder.sharesOwned.setText(String.format("%s shares owned", finalShrOwned.getQty()));
                    }
                }
            });

            while (true) {

                // Get market status
                String marketStatus = null;
                try {
                    marketStatus = polygonAPI.getMarketStatus().getMarket();

                } catch (PolygonAPIRequestException e) {
                    e.printStackTrace();
                }

                // Check if market is open
                if (marketStatus != null) {
                    if (marketStatus.equals("open")) {

                        // Get the last quote
                        LastQuoteResponse curr = null;
                        try {
                            curr = polygonAPI.getLastQuote(mData.get(position));

                        } catch (PolygonAPIRequestException e) {
                            e.printStackTrace();
                        }

                        try {
                            close = polygonAPI.getPreviousClose(mData.get(position), false).getResults().get(0).getC().floatValue();
                        } catch (PolygonAPIRequestException e) {
                            e.printStackTrace();
                        }

                        // Set values
                        LastQuoteResponse finalCurr = curr;
                        float finalClose = close;
                        mainActivity.runOnUiThread(() -> {

                            if (finalCurr != null && finalClose != 0) {
                                holder.priceOfStock.setText(String.format("$%.2f", finalCurr.getLast().getAskprice().floatValue()));
                                float temp = (finalCurr.getLast().getAskprice().floatValue() - finalClose) / finalClose * 100;

                                // Sets the precision and adds to view, then changes color
                                if (temp >= 0) {
                                    holder.percentChange.setText(String.format("+%.2f%%", temp));
                                    mClickListener.switchColors(holder, true);

                                } else {
                                    holder.percentChange.setText(String.format("%.2f%%", temp));
                                    mClickListener.switchColors(holder, false);
                                }

                            }
                        });

                    // Market isn't open
                    } else {

                        // Fetch Daily Open Close endpoint for prev day
                        // https://api.polygon.io/v1/open-close/AAPL/2020-10-14?apiKey=
                        /*JSONObject nodeHttpResponse = null;
                        try {
                            nodeHttpResponse = Unirest.get("https://api.polygon.io/v1/open-close/" + mData.get(position) + "/" + calendar.get(calendar.size() - 2).getDate() + "?apiKey=" + prefs.retrieveString("polygon_id", "NULL")).asJson().getBody().getObject();

                        } catch (UnirestException e) {
                            e.printStackTrace();
                        }*/

                        // Get last open day's close
    //                    if (nodeHttpResponse != null) {
                        try {
//                            closeCurr = polygonAPI.getLastQuote(mData.get(position)).getLast().getAskprice().floatValue();
                            ArrayList<Aggregate> aggs = polygonAPI.getAggregates(mData.get(position), 1, Timespan.DAY, lastOpenDate, lastOpenDate.plusDays(1), false).getResults();
                            closeCurr = aggs.get(aggs.size() - 1).getC().floatValue();
                        } catch (PolygonAPIRequestException e) {
                            e.printStackTrace();
                        }
    //                    }

                        // Set values
                        float finalClose = close;
                        float finalCurr = closeCurr;
                        mainActivity.runOnUiThread(() -> {

                            if (finalCurr != 0 && finalClose != 0) {
                                holder.priceOfStock.setText(String.format("$%.2f", finalCurr));
                                float temp = (finalCurr - finalClose) / finalClose * 100;

                                // Sets the precision and adds to view, then changes color
                                if (temp >= 0) {
                                    holder.percentChange.setText(String.format("+%.2f%%", temp));
                                    mClickListener.switchColors(holder, true);

                                } else {
                                    holder.percentChange.setText(String.format("%.2f%%", temp));
                                    mClickListener.switchColors(holder, false);
                                }

                            }
                        });

                    }
                }

                // Sleep for 1 minute
                try {
                    Thread.sleep(10000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        });
        thread.start();

    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    // total number of rows
    @Override
    public int getItemCount() {
        return mData.size();
    }


    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView stock_name;
        MaterialButton percentChange;
        TextView sharesOwned;
        TextView priceOfStock;
        MaterialCardView stockCard;

        ViewHolder(View itemView) {
            super(itemView);
            stock_name = itemView.findViewById(R.id.stockName);
            percentChange = itemView.findViewById(R.id.currentPrice);
            sharesOwned = itemView.findViewById(R.id.sharesOwned);
            priceOfStock = itemView.findViewById(R.id.priceOfTicker);
            stockCard = itemView.findViewById(R.id.positionCard);
            stockCard.setOnClickListener(this);
            itemView.setOnClickListener(this);
            percentChange.setOnClickListener(this);

        }

        @Override
        public void onClick(View view) {

            if (mClickListener != null) {
                mClickListener.onItemClick(view, getAdapterPosition());
            }
        }
    }

    // convenience method for getting data at click position
    String getItem(int id) {
        return mData.get(id);
    }

    // allows clicks events to be caught
    void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface ItemClickListener {
        void onItemClick(View view, int position);

        void switchColors(ViewHolder view, boolean pos);
    }
}

File: app/src/main/java/com/seanmiller/alpacadashboard/RecyclerViewAdapterPositions.kt
package com.seanmiller.alpacadashboard

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView
import net.jacobpeterson.alpaca.AlpacaAPI
import net.jacobpeterson.alpaca.model.endpoint.position.Position
import net.jacobpeterson.alpaca.model.properties.DataAPIType
import net.jacobpeterson.alpaca.model.properties.EndpointAPIType
import net.jacobpeterson.alpaca.rest.AlpacaClientException
//import net.jacobpeterson.alpaca.enums.api.DataAPIType
//import net.jacobpeterson.alpaca.enums.api.EndpointAPIType
//import net.jacobpeterson.alpaca.rest.exception.AlpacaAPIRequestException
//import net.jacobpeterson.domain.alpaca.position.Position
import java.text.DecimalFormat

class RecyclerViewAdapterPositions internal constructor(
    context: Context?,
    data: List<Position>,
    positionV: PositionView
) : RecyclerView.Adapter<RecyclerViewAdapterPositions.ViewHolder>() {
    private val mInflater: LayoutInflater = LayoutInflater.from(context)
    private var mClickListener: ItemClickListener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = mInflater.inflate(R.layout.recyclerview_position, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        // Set variables for each of the areas
        val mainActivity = MainActivity()
        val stockName = mData[position].symbol
        holder.stock_name.text = stockName
        val prefs = SharedPreferencesManager(mInflater.context)
        val alpacaAPI = AlpacaAPI(
            null,
            null,
            null,
            prefs!!.retrieveString("auth_token", "NULL"),
            EndpointAPIType.PAPER,
            DataAPIType.IEX
        )
        val alpacaData = AlpacaAPI(
            null,
            Properties.apiKey,
            Properties.secretKey,
            null,
            EndpointAPIType.PAPER,
            DataAPIType.IEX
        )

        holder.priceOfStock.text = String.format("$%.2f", mData[position].currentPrice.toFloat())

        val initialThread = Thread {

            // Initialize variables
            val percentChange = mData[position].changeToday
            val totalPercentChange = mData[position].unrealizedPlpc
            val totalReturn = mData[position].unrealizedPl
            if (percentChange != "-1" || positionView != PositionView.PERCENT_CHANGE) {
                // Update the "percent change" parameter in holder
                mainActivity.runOnUiThread {
                    updateValues(
                        holder,
                        percentChange,
                        totalPercentChange,
                        totalReturn
                    )
                }

            } else {
                // Set values
                try {
                    val snapshot = alpacaData.marketData().getSnapshot(stockName)
                    val finalClose = snapshot.prevDailyBar.c.toFloat()
//                      val finalCurr = snapshot.latestQuote.ap.toFloat()
                    val finalCurr = snapshot.dailyBar.c.toFloat()

                    mainActivity.runOnUiThread {
                        if (finalCurr != 0f && finalClose != 0f) {
                            holder.priceOfStock.text = String.format("$%.2f", finalCurr)
                            val temp = (finalCurr - finalClose) / finalClose

                            updateValues(holder, temp.toString(), totalPercentChange, totalReturn)
                        }
                    }

                } catch (e: AlpacaClientException) {
                    e.printStackTrace()
                }

            }
        }
        initialThread.start()

        val updateThread = Thread {

            while (true) {

                // Get Amount of shares owned
                var shrOwned: Position? = null
                try {
                    shrOwned = alpacaAPI.positions().getBySymbol(stockName)
                } catch (e: AlpacaClientException) {
                    e.printStackTrace()
                }

                // Set shares owned
                val finalShrOwned = shrOwned
                mainActivity.runOnUiThread {
                    if (finalShrOwned == null) {
                        holder.sharesOwned.text = String.format("%s shares owned", 0)
                        return@runOnUiThread;
                    }
                    when (finalShrOwned.qty) {
                        "1" -> holder.sharesOwned.text =
                            String.format("%s share owned", finalShrOwned.qty)

                        null -> holder.sharesOwned.text = String.format("%s shares owned", 0)

                        else -> holder.sharesOwned.text =
                            String.format("%s shares owned", finalShrOwned.qty)

                    }
                }

                // Sleep for 1 minute
                try {
                    Thread.sleep(60000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

                // Set values
                try {
                    val pos: Position = alpacaAPI.positions().getBySymbol(stockName)
                    val percentChange = pos.changeToday
                    val totalPercentChange = pos.unrealizedPlpc
                    val totalReturn = pos.unrealizedPl
                    if (mData[position].changeToday != "-1") { // If market day
                        mainActivity.runOnUiThread {
                            updateValues(
                                holder,
                                percentChange,
                                totalPercentChange,
                                totalReturn
                            )
                        }

                    } else {
                        val snapshot = alpacaData.marketData().getSnapshot(stockName)
                        val finalClose = snapshot.prevDailyBar.c.toFloat()
//                      val finalCurr = snapshot.latestQuote.ap.toFloat()
                        val finalCurr = snapshot.dailyBar.c.toFloat()

                        mainActivity.runOnUiThread {
                            if (finalCurr != 0f && finalClose != 0f) {
                                holder.priceOfStock.text = String.format("$%.2f", finalCurr)
                                val temp = (finalCurr - finalClose) / finalClose

                                updateValues(
                                    holder,
                                    temp.toString(),
                                    totalPercentChange,
                                    totalReturn
                                )
                            }
                        }
                    }

                } catch (e: AlpacaClientException) {
                    e.printStackTrace()
                }

            }
        }
        updateThread.start()
    }

    private fun updateValues(
        holder: ViewHolder,
        percentChange: String,
        totalPercentChange: String,
        totalReturn: String
    ) {

        // Sets value based on what positionView is
        if (positionView == PositionView.PERCENT_CHANGE) {
            // Sets the precision and adds to view, then changes color
            if (percentChange.toFloat() >= 0) {
                holder.percentChange.text = String.format("+%.2f%%", percentChange.toFloat() * 100)
                mClickListener!!.switchColors(holder, true)
            } else {
                holder.percentChange.text = String.format("%.2f%%", percentChange.toFloat() * 100)
                mClickListener!!.switchColors(holder, false)
            }

        } else if (positionView == PositionView.TOTAL_PERCENT_CHANGE) {
            // Sets the precision and adds to view, then changes color
            if (totalPercentChange.toFloat() >= 0) {
                holder.percentChange.text =
                    String.format("+%.2f%%", totalPercentChange.toFloat() * 100)
                mClickListener!!.switchColors(holder, true)
            } else {
                holder.percentChange.text =
                    String.format("%.2f%%", totalPercentChange.toFloat() * 100)
                mClickListener!!.switchColors(holder, false)
            }

        } else if (positionView == PositionView.TOTAL_RETURN) {
            // Sets the precision and adds to view, then changes color
            val formatter = DecimalFormat("#,###")
            val amount =
                "$" + formatter.format(kotlin.math.abs(totalReturn.toFloat() * 100)).toString()
            if (totalReturn.toFloat() >= 0) {
                holder.percentChange.text = String.format("+%s", amount)
                mClickListener!!.switchColors(holder, true)
            } else {
                holder.percentChange.text = String.format("-%s", amount)
                mClickListener!!.switchColors(holder, false)
            }

        }
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    // total number of rows
    override fun getItemCount(): Int {
        return mData.size
    }

    // stores and recycles views as they are scrolled off screen
    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        var stock_name: TextView = itemView.findViewById(R.id.stockName)
        var percentChange: MaterialButton = itemView.findViewById(R.id.currentPrice)
        var sharesOwned: TextView = itemView.findViewById(R.id.sharesOwned)
        var priceOfStock: TextView = itemView.findViewById(R.id.priceOfTicker)
        var stockCard: MaterialCardView = itemView.findViewById(R.id.positionCard)
        override fun onClick(view: View) {
            if (mClickListener != null) {
                mClickListener!!.onItemClick(view, adapterPosition)
            }
        }

        init {
            stockCard.setOnClickListener(this)
            itemView.setOnClickListener(this)
            percentChange.setOnClickListener(this)
        }
    }

    // convenience method for getting data at click position
    fun getItem(id: Int): Position {
        return mData[id]
    }

    // allows clicks events to be caught
    fun setClickListener(itemClickListener: ItemClickListener?) {
        mClickListener = itemClickListener
    }

    fun notifyItemChanged(positionV: PositionView?) {
        positionView = positionV!!
    }

    // parent activity will implement this method to respond to click events
    interface ItemClickListener {
        fun onItemClick(view: View?, position: Int)
        fun switchColors(view: ViewHolder?, pos: Boolean)
    }

    companion object {
        private lateinit var mData: List<Position>
        private lateinit var positionView: PositionView
    }

    // data is passed into the constructor
    init {
        mData = data
        positionView = positionV
    }
}